### DISCUSSION TITLE
Support stream Chunk reading in Python API

### DISCUSSION THREAD

**Discussion by watwea-heaviside** [foxglove/mcap#1550] [General]
In the Python mcap reader API, chunks are read into memory as full blocks. This often causes OOM errors on smaller systems.

Main reader loop: https://github.com/foxglove/mcap/blob/a68c76979d646fa90a24bff3457be2da0c371f0a/python/mcap/mcap/reader.py#L294-L318
Chunk data read line: https://github.com/foxglove/mcap/blob/0a06331980a47ef606cca06b6ee3fc987f5f2d52/python/mcap/mcap/records.py#L173-L173

**Comment by wkalt** [comment]
I think this idea is good generally. The main constraint is that to make use of the message index, you need to decompress the whole chunk first. But if you don't care about the message index and you're just doing a linear file read, then streaming decompression of chunks is what you want for sure. We support this in the go library but none of the others afaik.

I think the main place where you benefit from this is when reading a lot of mcap files at once (for instance, merging them together). OOMs during reading of a single MCAP file may also be addressed, but not if they are caused by oversized individual messages, which is the most common reason for oversized chunks in my experience. Absent massive messages, if the writer is configured with a sane chunk size then OOMs should not generally be encountered when reading a single file -- the default decompressed chunk sizes are in the single-digit megabytes, and you generally only need one at a time open.

So to summarize, I think the feature is good but it would be worth checking if there's anything odd going on with your chunking, if the context here is reading single files. You can do this with

```
mcap info file.mcap
```

and 

```
mcap list chunks file.mcap
```

which will show all the individual sizes. There was an old bug in the rust writer that resulted in single-chunk files. I'm curious if you may be on an old version of the rust writer that still has this behavior: 
https://github.com/foxglove/mcap/pull/777

or are creating huge chunks in some other way.
